lua/cmp/entry.lua

```diff
@@ -204,244 +261,233 @@ end
-  else
-    insert_range = {
-      start = {
-        line = self.context.cursor.row - 1,
-        character = self:get_offset() - 1,
-      },
-      ['end'] = self.source_insert_range['end'],
-    }
-  end
-  return insert_range
+  return self.insert_range
 end
 
----Return replace range
----@return lsp.Range|nil
+---@deprecated use entry.replace_range instead
 entry.get_replace_range = function(self)
-  return self.cache:ensure('get_replace_range', function()
-    local replace_range
-    if self:get_completion_item().textEdit then
-      if self:get_completion_item().textEdit.replace then
-        replace_range = self:get_completion_item().textEdit.replace
-      else
-        replace_range = self:get_completion_item().textEdit.range --[[@as lsp.Range]]
-      end
-      replace_range = self:convert_range_encoding(replace_range)
-    end
-
-    if not replace_range or ((self.context.cursor.col - 1) == replace_range['end'].character) then
-      replace_range = {
-        start = {
-          line = self.source_replace_range.start.line,
-          character = self:get_offset() - 1,
-        },
-        ['end'] = self.source_replace_range['end'],
-      }
-    end
-    return replace_range
-  end)
+  return self.replace_range
 end
 
 ---Match line.
 ---@param input string
 ---@param matching_config cmp.MatchingConfig
 ---@return { score: integer, matches: table[] }
 entry.match = function(self, input, matching_config)
-  return self.match_cache:ensure(input .. ':' .. (self.resolved_completion_item and '1' or '0' .. ':') .. (matching_config.disallow_fuzzy_matching and '1' or '0') .. ':' .. (matching_config.disallow_partial_fuzzy_matching and '1' or '0') .. ':' .. (matching_config.disallow_partial_matching and '1' or '0') .. ':' .. (matching_config.disallow_prefix_unmatching and '1' or '0') .. ':' .. (matching_config.disallow_symbol_nonprefix_matching and '1' or '0'), function()
-    local option = {
-      disallow_fuzzy_matching = matching_config.disallow_fuzzy_matching,
-      disallow_partial_fuzzy_matching = matching_config.disallow_partial_fuzzy_matching,
-      disallow_partial_matching = matching_config.disallow_partial_matching,
-      disallow_prefix_unmatching = matching_config.disallow_prefix_unmatching,
-      disallow_symbol_nonprefix_matching = matching_config.disallow_symbol_nonprefix_matching,
-      synonyms = {
-        self:get_word(),
-        self:get_completion_item().label,
-      },
-    }
+  -- https://www.lua.org/pil/11.6.html
+  -- do not use '..' to allocate multiple strings
+  local cache_key = string.format('%s:%d:%d:%d:%d:%d:%d', input, self.resolved_completion_item and 1 or 0, matching_config.disallow_fuzzy_matching and 1 or 0, matching_config.disallow_partial_matching and 1 or 0, matching_config.disallow_prefix_unmatching and 1 or 0, matching_config.disallow_partial_fuzzy_matching and 1 or 0, matching_config.disallow_symbol_nonprefix_matching and 1 or 0)
```

Comment:
In my opinion this does not affect performance.
The issue of allocating multiple strings only arises when concatenating strings in a loop.

(But I can accept this.)

Reply:
My benchmark is here.

```lua
local ITERATION = 1000000000

local function consume(_)
end

local function run(name, f)
  collectgarbage('collect')
  collectgarbage('setpause')
  local s = os.clock()
  for _ = 0, ITERATION, 1 do
    consume(f())
  end
  local e = os.clock()
  print(('%s'):format(name))
  print(('    %s seconds'):format(e - s))
  print(('    %s kb'):format(collectgarbage('count')))
end

local a = string.rep('a', 100)
local b = string.rep('b', 100)
local c = string.rep('c', 100)
local d = string.rep('d', 100)
local e = string.rep('e', 100)
local f = string.rep('f', 100)
local g = string.rep('g', 100)

vim.print(('Lua version %s (%s)'):format(_VERSION, type(jit) == 'table' and 'JIT' or 'PUC'))

for i = 1, 10 do
  run(('[%s] format'):format(i), function()
    return string.format('%s%s%s%s%s%s%s', a, b, c, d, e, f, g)
  end)
  run(('[%s] syntax'):format(i), function()
    return a .. b .. c .. d .. e .. f .. g
  end)
end
```

```
Lua version Lua 5.1 (JIT)
[1] format
    0.309662 seconds
    7442.2548828125 kb
[1] syntax
    0.287758 seconds
    7442.9619140625 kb
[2] format
    0.288337 seconds
    7443.9189453125 kb
[2] syntax
    0.289305 seconds
    7444.6806640625 kb
[3] format
    0.28824099999999 seconds
    7445.8564453125 kb
[3] syntax
    0.288509 seconds
    7447.3330078125 kb
[4] format
    0.28775099999999 seconds
    7446.8486328125 kb
[4] syntax
    0.29039900000001 seconds
    7446.5986328125 kb
[5] format
    0.288085 seconds
    7446.4580078125 kb
[5] syntax
    0.290503 seconds
    7446.3955078125 kb
[6] format
    0.28749599999998 seconds
    7446.4111328125 kb
[6] syntax
    0.287892 seconds
    7446.3955078125 kb
[7] format
    0.28724800000001 seconds
    7446.4111328125 kb
[7] syntax
    0.287656 seconds
    7446.3955078125 kb
[8] format
    0.28767599999998 seconds
    7446.4111328125 kb
[8] syntax
    0.28742099999999 seconds
    7446.4111328125 kb
[9] format
    0.28775399999998 seconds
    7446.4111328125 kb
[9] syntax
    0.28724700000001 seconds
    7446.4111328125 kb
[10] format
    0.28726999999998 seconds
    7446.4111328125 kb
[10] syntax
    0.287116 seconds
    7446.3955078125 kb
```

Reply:
Can you try with nested expressions (using `()`)?

Reply:
Benchmark updated.

However, it was necessary to use math.random.
This may have become a benchmark for math.random.

```lua
local ITERATION = 10000000

local function consume(_)
end

local function run(name, f)
  collectgarbage('collect')
  collectgarbage('stop')
  local s = os.clock()
  for _ = 0, ITERATION, 1 do
    consume(f())
  end
  local e = os.clock()
  print(('%s'):format(name))
  print(('    %s seconds'):format(e - s))
  print(('    %s kb'):format(collectgarbage('count')))
  collectgarbage('restart')
end


vim.print(('Lua version %s (%s)'):format(_VERSION, type(jit) == 'table' and 'JIT' or 'PUC'))

for i = 1, 10 do
  run(('[%s] format'):format(i), function()
    local a = tostring(math.random())
    local b = tostring(math.random())
    local c = tostring(math.random())
    return string.format('%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s', a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c)
  end)
  run(('[%s] syntax'):format(i), function()
    local a = tostring(math.random())
    local b = tostring(math.random())
    local c = tostring(math.random())
    return (a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c)
  end)
end
```


```
Lua version Lua 5.1 (JIT)
[1] format
    0.173224 seconds
    7593.55078125 kb
[1] syntax
    0.161859 seconds
    6650.484375 kb
[2] format
    0.158104 seconds
    6652.5 kb
[2] syntax
    0.159953 seconds
    6654.12890625 kb
[3] format
    0.158455 seconds
    6656.18359375 kb
[3] syntax
    0.157741 seconds
    6656.9921875 kb
[4] format
    0.160288 seconds
    6645.15234375 kb
[4] syntax
    0.160823 seconds
    6644.81640625 kb
[5] format
    0.161516 seconds
    6644.77734375 kb
[5] syntax
    0.159043 seconds
    6644.62890625 kb
[6] format
    0.159526 seconds
    6644.71484375 kb
[6] syntax
    0.159805 seconds
    6644.62890625 kb
[7] format
    0.160698 seconds
    6644.71484375 kb
[7] syntax
    0.159825 seconds
    6644.62890625 kb
[8] format
    0.160317 seconds
    6644.71484375 kb
[8] syntax
    0.161488 seconds
    6644.62890625 kb
[9] format
    0.159925 seconds
    6644.71484375 kb
[9] syntax
    0.159897 seconds
    6644.62890625 kb
[10] format
    0.15959 seconds
    6644.70703125 kb
[10] syntax
    0.159551 seconds
    6644.62890625 kb```

Reply:
The below version's result is here.

```
Lua version Lua 5.1 (JIT)
[1] format
    0.178772 seconds
    5814.7900390625 kb
[1] syntax
    0.15985 seconds
    5828.2861328125 kb
[2] format
    0.161948 seconds
    5821.6298828125 kb
[2] syntax
    0.160109 seconds
    5831.9501953125 kb
[3] format
    0.159681 seconds
    5824.6220703125 kb
[3] syntax
    0.16077 seconds
    5834.7314453125 kb
[4] format
    0.161918 seconds
    5814.1650390625 kb
[4] syntax
    0.159785 seconds
    5813.8291015625 kb
[5] format
    0.159689 seconds
    5813.7861328125 kb
[5] syntax
    0.159928 seconds
    5813.6416015625 kb
[6] format
    0.159769 seconds
    5813.7275390625 kb
[6] syntax
    0.160404 seconds
    5813.6416015625 kb
[7] format
    0.161152 seconds
    5813.7275390625 kb
[7] syntax
    0.160442 seconds
    5813.6416015625 kb
[8] format
    0.160327 seconds
    5813.7275390625 kb
[8] syntax
    0.160148 seconds
    5813.6416015625 kb
[9] format
    0.160261 seconds
    5813.7275390625 kb
[9] syntax
    0.159532 seconds
    5813.6416015625 kb
[10] format
    0.159259 seconds
    5813.7275390625 kb
[10] syntax
    0.161971 seconds
    5813.6416015625 kb
```

It seems there is no difference.

Reply:
Good to know. Thanks again for doing that. I'll reference this for future discussions.

Reply:
Sorry. my benchmark is wrong.

It seems to affect to performance to use `string.format` instead.

```lua
local ITERATION = 1000000

local function consume(data)
  _G.data = data
end

local function run(name, f)
  collectgarbage('collect')
  collectgarbage('stop')
  local s = os.clock()
  for _ = 0, ITERATION, 1 do
    consume(f())
  end
  local e = os.clock()
  print(('%s'):format(name))
  print(('    %s seconds'):format(e - s))
  print(('    %s kb'):format(collectgarbage('count')))
  collectgarbage('restart')
end


vim.print(('Lua version %s (%s)'):format(_VERSION, type(jit) == 'table' and 'JIT' or 'PUC'))

for i = 1, 10 do
  run(('[%s] format'):format(i), function()
    local a = tostring(math.random())
    local b = tostring(math.random())
    local c = tostring(math.random())
    return string.format('%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s', a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c, a, b, c)
  end)
  run(('[%s] syntax'):format(i), function()
    local a = tostring(math.random())
    local b = tostring(math.random())
    local c = tostring(math.random())
    return (a .. b .. c .. a .. b .. c .. a .. b) .. (c .. (a .. b) .. (c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c .. a .. b .. c) .. a .. b .. c .. (a .. b .. c .. a .. b .. c) .. a .. b .. (c .. a .. b .. c) .. a .. b .. c)
  end)
end

```


```
Lua version Lua 5.1 (JIT)
[1] format
    2.233753 seconds
    1088391.6679688 kb
[1] syntax
    3.371306 seconds
    2050993.4609375 kb
[2] format
    1.949485 seconds
    1088394.8554688 kb
[2] syntax
    3.23479 seconds
    2051043.8085938 kb
[3] format
    2.216916 seconds
    1088433.4296875 kb
[3] syntax
    3.398898 seconds
    2051061.4921875 kb
[4] format
    1.848707 seconds
    1088388.6757813 kb
[4] syntax
    3.150517 seconds
    2051042.03125 kb
[5] format
    1.838001 seconds
    1088439.6914063 kb
[5] syntax
    3.136715 seconds
    2051044.0859375 kb
[6] format
    1.865727 seconds
    1088417.75 kb
[6] syntax
    3.187182 seconds
    2051002.7890625 kb
[7] format
    1.915407 seconds
    1088402.2890625 kb
[7] syntax
    3.131942 seconds
    2051061.3242188 kb
[8] format
    1.882645 seconds
    1088402.6679688 kb
[8] syntax
    3.171707 seconds
    2051076.8320313 kb
[9] format
    1.829586 seconds
    1088384.7734375 kb
[9] syntax
    3.060516 seconds
    2050994.234375 kb
[10] format
    1.831612 seconds
    1088394.1328125 kb
[10] syntax
    3.126076 seconds
    2051067.203125 kb
```

Reply:
Sorry, do you mean `string.format` was always faster? What specifically is different in the last run?

Reply:
The difference is 
```
local function consume(data)
  _G.data = data -- ← ADD THIS
end
```

Probably, this disables JIT compilation, which makes it slower for `syntax` case.

The results show that `string.format` is consistently faster.
And there's no speed difference in a JIT environment.